### PR TITLE
 Fix #2143: Create a unique link for each of the entry on the leaderboard

### LINKED ISSUE #2143: Create a unique link for each of the entry on the leaderboard
## Current Scenario

If a participant wants to share their rank in a challenge on twitter/facebook then they have to share the entire leaderboard link.

## Expected scenario
There should be a unique link for each entry and upon opening, it should scroll to that particular entry and highlight it.

### PR DESCRIPTION
**Pull request by Sanji515**
**Deliverables:**

- [x] Used `$anchorScroll` service to scroll to a specified element in a page
- [x] Instead of giving unique link on clicking rank, given on entire `tr` tag
- [x] Added `css` for highlighting the entry when someone tries to open link

### CODE REVIEW

**Comment by Sanji515** [comment]
@RishabhJain2018 @galipremsagar Please review

**Comment by galipremsagar** [comment]
Cool, Will review this today.

File: frontend/src/js/controllers/challengeCtrl.js [hunk 1]
```diff
@@ -62,6 +62,30 @@
...
+        
+        // get unique rank number from the url & if exists hightlight the entry
+        var elementId = $location.absUrl().split('?')[0].split('#')[1];
+        if (elementId) {
+            $timeout(function () {
+                $anchorScroll(elementId);
+                $scope.currentEntryLink = elementId;
+                $anchorScroll.yOffset = 90;
+                $scope.isHighlight = elementId.split("leaderboardrank-")[1];
+            }, 2000);
```
Comment (galipremsagar) [inline frontend/src/js/controllers/challengeCtrl.js]: Why is this too in a timeout??

Comment (galipremsagar) [reply frontend/src/js/controllers/challengeCtrl.js]: Remove this from timeout and trigger this function once the leaderboard is loaded i.e., in on-success

Comment (Sanji515) [reply frontend/src/js/controllers/challengeCtrl.js]: @galipremsagar hey, as it's still not working after moving to the `get leaderboard` function so I moved the code to the polling section of leaderboard as it is sure that polling will start after the leaderboard is loaded and then called my code by putting inside a function only once when polling starts, please check if it's valid and working on your side

File: frontend/src/js/controllers/challengeCtrl.js [hunk 2]
```diff
@@ -50,6 +50,7 @@
         vm.loaderTitle = '';
         vm.loaderContainer = angular.element('.exist-team-card');
         vm.termsAndConditions = false;
+        vm.isLeaderboardLoaded = false;
```
Comment (galipremsagar) [inline frontend/src/js/controllers/challengeCtrl.js]:
```suggestion
        vm.isScrollComplete = false;
```

File: frontend/src/js/controllers/challengeCtrl.js [hunk 3]
```diff
@@ -653,9 +678,13 @@
             };
 
             utilities.sendRequest(parameters);
-            vm.startLeaderboard = function() {
-                vm.stopLeaderboard();
+            vm.startLeaderboard = function() {             
+                vm.stopLeaderboard();
```
Comment (galipremsagar) [inline frontend/src/js/controllers/challengeCtrl.js]:
```suggestion
                vm.stopLeaderboard();           
                if (vm.isScrollComplete === false) {
                    vm.scrollToEntryAfterLeaderboardLoads();
                    vm.isScrollComplete = true;
                }
```

File: frontend/src/js/controllers/challengeCtrl.js [hunk 4]
```diff
@@ -653,9 +678,13 @@
             };
 
             utilities.sendRequest(parameters);
-            vm.startLeaderboard = function() {
-                vm.stopLeaderboard();
+            vm.startLeaderboard = function() {             
+                vm.stopLeaderboard();           
                 vm.poller = $interval(function() {
+                    if (vm.isLeaderboardLoaded === false) {
```
Comment (galipremsagar) [inline frontend/src/js/controllers/challengeCtrl.js]:
```suggestion
```

File: frontend/src/js/controllers/challengeCtrl.js [hunk 5]
```diff
@@ -653,9 +678,13 @@
...
+                        vm.scrollToEntryAfterLeaderboardLoads();
```
Comment (galipremsagar) [inline frontend/src/js/controllers/challengeCtrl.js]:
```suggestion
```

File: frontend/src/js/controllers/challengeCtrl.js [hunk 6]
```diff
@@ -653,9 +678,13 @@
...
+                    }
```
Comment (galipremsagar) [inline frontend/src/js/controllers/challengeCtrl.js]:
```suggestion
```

File: frontend/src/js/controllers/challengeCtrl.js [hunk 7]
```diff
@@ -653,9 +678,13 @@
...
+                    vm.isLeaderboardLoaded = true;
```
Comment (galipremsagar) [inline frontend/src/js/controllers/challengeCtrl.js]:
```suggestion
```

File: frontend/src/views/web/challenge/leaderboard.html [hunk 8]
```diff
@@ -62,7 +62,7 @@
                                     <thead>
                                         <tr>
                                             <td data-field="rank" class="align-center">
-                                                <a href="#" ng-click="challenge.sortLeaderboard(challenge, 'rank');">
+                                                <a href="#{{currentEntryLink}}" ng-click="challenge.sortLeaderboard(challenge, 'rank');">
```
Comment (galipremsagar) [inline frontend/src/views/web/challenge/leaderboard.html]:
```suggestion
                                                <a href="#" ng-click="challenge.sortLeaderboard(challenge, 'rank');">
```

File: frontend/src/views/web/challenge/leaderboard.html [hunk 9]
```diff
@@ -73,7 +73,7 @@
                                                 </a>
                                             </td>
                                             <td data-field="team" class="align-center">
-                                                <a href="#" ng-click="challenge.sortLeaderboard(challenge, 'string');">
+                                                <a href="#{{currentEntryLink}}" ng-click="challenge.sortLeaderboard(challenge, 'string');">
```
Comment (galipremsagar) [inline frontend/src/views/web/challenge/leaderboard.html]:
```suggestion
                                                <a href="#" ng-click="challenge.sortLeaderboard(challenge, 'string');">
```
Revert this

File: frontend/src/views/web/challenge/leaderboard.html [hunk 10]
```diff
@@ -85,7 +85,7 @@
                                             </td>
                                             <td ng-repeat="key in challenge.leaderboard[0].leaderboard__schema.labels"
                                                 class="align-center">
-                                                <a href="#"
+                                                <a href="#{{currentEntryLink}}"
```
Comment (galipremsagar) [inline frontend/src/views/web/challenge/leaderboard.html]:
```suggestion
                                                <a href="#"
```
revert this

File: frontend/src/views/web/challenge/leaderboard.html [hunk 11]
```diff
@@ -97,7 +97,7 @@
                                                 </a>
                                             </td>
                                             <td data-field="submission_time" class="align-center">
-                                                <a href="#" ng-click="challenge.sortLeaderboard(challenge, 'date');">
+                                                <a href="#{{currentEntryLink}}" ng-click="challenge.sortLeaderboard(challenge, 'date');">
```
Comment (galipremsagar) [inline frontend/src/views/web/challenge/leaderboard.html]:
```suggestion
                                                <a href="#" ng-click="challenge.sortLeaderboard(challenge, 'date');">
```
revert this

File: frontend/src/js/controllers/challengeCtrl.js [hunk 12]
```diff
@@ -63,6 +64,30 @@
...
+            var newHash = elementId.toString();
+            if ($location.hash() !== newHash) {
+                $location.hash(elementId);
+                $scope.currentEntryLink = elementId;
+            } else {
+                $anchorScroll();
+            }
+            $scope.isHighlight = false;
+            $anchorScroll.yOffset = 90;
+        };
```
Comment (galipremsagar) [inline frontend/src/js/controllers/challengeCtrl.js]: Change this function to this:
```
vm.scrollToEntryAfterLeaderboardLoads = function () {
            // get unique rank number from the url & if exists hightlight the entry
            $timeout(function() {
                const elementId = $location.absUrl().split('?')[0].split('#')[1];
                console.log(elementId);
                if (elementId) {
                    $anchorScroll.yOffset = 90;
                    $anchorScroll(elementId);
                    $scope.currentEntryLink = elementId;
                    $scope.isHighlight = elementId.split("leaderboardrank-")[1];
                }
            }, 500);
        };
```

This is due this issue we are having: https://stackoverflow.com/a/30247770/4415561
and in this case it's okay to wait 500ms

**Pull request merged**